crates/wit-dylib/wit_dylib.h

```diff
@@ -0,0 +1,322 @@
+} wit_list_t;
+
+typedef struct wit_fixed_size_list {
+     const char *interface;
+     const char *name;
+     size_t size;
+     wit_type_t ty;
+} wit_fixed_size_list_t;
+
+typedef struct wit_future {
+     const char *interface;
+     const char *name;
+     wit_type_t ty;
+} wit_future_t;
+
+typedef struct wit_stream {
+     const char *interface;
+     const char *name;
+     wit_type_t ty;
+} wit_stream_t;
+
+typedef struct wit_alias {
+     const char *interface;
+     const char *name;
+     wit_type_t ty;
+} wit_alias_t;
+
+#define WIT_V0 0
+
+typedef struct wit {
+     uint32_t version; // `WIT_V*`
+
+     size_t num_funcs;
+     const wit_func_t *funcs;
+     size_t num_resources;
+     const wit_resource_t *resources;
+     size_t num_records;
+     const wit_record_t *records;
+     size_t num_flags;
+     const wit_flags_t *flags;
+     size_t num_tuples;
+     const wit_tuple_t *tuples;
+     size_t num_variants;
+     const wit_variant_t *variants;
+     size_t num_enums;
+     const wit_enum_t *enums;
+     size_t num_options;
+     const wit_option_t *options;
+     size_t num_results;
+     const wit_result_t *results;
+     size_t num_lists;
+     const wit_list_t *lists;
+     size_t num_fixed_size_lists;
+     const wit_fixed_size_list_t *fixed_size_lists;
+     size_t num_futures;
+     const wit_future_t *futures;
+     size_t num_streams;
+     const wit_stream_t *streams;
+     size_t num_aliases;
+     const wit_alias_t *aliases;
+} wit_t;
+
+// Invoked during `__wasm_call_ctors` with an in-memory `wit_t` data structure.
+//
+// The pointer provided lives for the lifetime of the entire program so it's
+// safe to store this pointer.
+void wit_dylib_initialize(const wit_t* wit);
+
+// Entrypoint for WIT exports.
+//
+// `which` is an index into `wit->funcs` and `ptr` contains the lifted
+// parameters of the functionc all. The result, if any, is stored at `ptr`.
+void wit_dylib_call_export(size_t which, uint64_t *ptr);
+
+// Entrypoint for WIT resource destructors.
+//
+// The `ty` poitns to `wit->resources` and `handle` is the value being
+// destroyed.
+void wit_dylib_resource_dtor(size_t ty, size_t handle);
+
+// Generic byte deallocation function.
+void wit_dylib_dealloc_bytes(void *ptr, size_t byte_size, size_t align);
+
+// Deallocates an interpreter `val` provided, if necessary.
+void wit_dylib_dealloc_val(uint64_t val);
+
+// =============================================================================
+// Functions used to lower values, or convert from interpreter `uint64_t`
+// representations to WIT typed representations.
+//
+// Note that during lowering a `uint64_t` is NOT an "owned" value meaning that
+// these functions should not allocate memory as it otherwise won't get cleaned
+// up. For example the return value of `wit_dylib_list_get` is considered to be
+// "borrowed" and not needing cleanup. This is suitable if, for example, the
+// returned value is a pointer into the original list.
```

Comment:
I think it'd be good to expand on this comment a bit / point to some other source of documentation of how this is supposed to be used. Specifically, from this it's unclear what an implementer is supposed to do if their value representation doesn't allow them to follow the "don't allocate" restriction here.

Additionally, IIUC this applies to `uint64_t *` out params as well, right? And of course to returned const pointers, though making that explicit would be good, too.